pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,385 @@
+		if p.defaultCacheDuration == 0 {
+			return dockerConfig
+		}
+
+		expiresAt = time.Now().Add(p.defaultCacheDuration)
+	} else {
+		expiresAt = time.Now().Add(response.CacheDuration.Duration)
+	}
+
+	cachedEntry := &cacheEntry{
+		key:         cacheKey,
+		credentials: dockerConfig,
+		expiresAt:   expiresAt,
+	}
+
+	if err := p.cache.Add(cachedEntry); err != nil {
+		klog.Errorf("Error adding auth entry to cache: %v", err)
+	}
+
+	return dockerConfig
+}
+
+// Enabled always returns true since registration of the plugin via kubelet implies it should be enabled.
+func (e *pluginProvider) Enabled() bool {
+	return true
+}
+
+// isImageAllowed returns true if the image matches against the list of allowed matches by the plugin.
+func (p *pluginProvider) isImageAllowed(image string) bool {
+	for _, matchImage := range p.matchImages {
+		if matched, _ := credentialprovider.URLsMatchStr(matchImage, image); matched {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getCachedCredentials returns a credentialprovider.DockerConfig if cached from the plugin.
+func (p *pluginProvider) getCachedCredentials(image string) (credentialprovider.DockerConfig, bool, error) {
+	obj, found, err := p.cache.GetByKey(image)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	registry := parseRegistry(image)
+	obj, found, err = p.cache.GetByKey(registry)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	obj, found, err = p.cache.GetByKey(globalCacheKey)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if found {
+		return obj.(*cacheEntry).credentials, true, nil
+	}
+
+	return nil, false, nil
+}
+
+// Plugin is the interface calling ExecPlugin. This is mainly for testability
+// so tests don't have to actually exec any processes.
+type Plugin interface {
+	ExecPlugin(ctx context.Context, image string) (*credentials.CredentialProviderResponse, error)
+}
+
+// execPlugin is the implementation of the Plugin interface that execs a credential provider plugin based
+// on it's name provided in CredentialProviderConfig. It is assumed that the executable is available in the
+// plugin directory provided by the kubelet.
+type execPlugin struct {
+	name         string
+	apiVersion   string
+	args         []string
+	envVars      []kubeletconfig.ExecEnvVar
+	pluginBinDir string
+}
+
+// ExecPlugin executes the plugin binary with arguments and environment variables specified in CredentialProviderConfig:
+//
+//  $ ENV_NAME=ENV_VALUE <plugin-name> args[0] args[1] <<<request
+//
+// The plugin is expected to receive the CredentialProviderRequest API via stdin from the kubelet and
+// return CredentialProviderResponse via stdout.
+func (e *execPlugin) ExecPlugin(ctx context.Context, image string) (*credentials.CredentialProviderResponse, error) {
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	stdin := &bytes.Buffer{}
```

Comment:
`stdin := bytes.NewBuffer(data)` instead of initializing it empty and then writing to it?